zebra-network/src/peer_set/initialize.rs

```diff
@@ -881,17 +921,79 @@ where
     }
 }
 
+/// Try to get more peers using `candidates`, then queue a connection attempt using `demand_tx`.
+/// If there were no new peers, the connection attempt is skipped.
+#[instrument(skip(candidates, demand_tx))]
+async fn crawl<S>(
+    candidates: Arc<futures::lock::Mutex<CandidateSet<S>>>,
+    mut demand_tx: futures::channel::mpsc::Sender<MorePeers>,
+) -> Result<(), BoxError>
+where
+    S: Service<Request, Response = Response, Error = BoxError> + Send + Sync + 'static,
+    S::Future: Send + 'static,
+{
+    // update() has timeouts, and briefly holds the address book
+    // lock, so it shouldn't hang.
+    // Try to get new peers, holding the lock for as short a time as possible.
+    let result = {
+        let result = candidates.lock().await.update().await;
+        std::mem::drop(candidates);
+        result
+    };
+    let more_peers = match result {
+        Ok(more_peers) => more_peers,
+        Err(e) => {
+            info!(
+                ?e,
+                "candidate set returned an error, is Zebra shutting down?"
+            );
+            return Err(e);
+        }
+    };
+
+    // If we got more peers, try to connect to a new peer on our next loop.
+    //
+    // # Security
+    //
+    // Update attempts are rate-limited by the candidate set,
+    // and we only try peers if there was actually an update.
+    //
+    // So if all peers have had a recent attempt, and there was recent update
+    // with no peers, the channel will drain. This prevents useless update attempt
+    // loops.
+    if let Some(more_peers) = more_peers {
+        if let Err(send_error) = demand_tx.try_send(more_peers) {
+            if send_error.is_disconnected() {
+                // Zebra is shutting down
+                return Err(send_error.into());
+            }
+        }
+    }
+
+    Ok(())
+}
```

Comment:
```suggestion
    Ok(try_send_demand(more_peers, &mut demand_tx)?)
}

fn try_send_demand(
    more_peers: Option<MorePeers>,
    demand_tx: &mut futures::channel::mpsc::Sender<MorePeers>,
) -> Result<(), futures::channel::mpsc::TrySendError<MorePeers>> {
    more_peers
        .and_then(|more_peers: MorePeers| demand_tx.try_send(more_peers).err())
        .filter(futures::channel::mpsc::TrySendError::is_disconnected) // Zebra is shutting down?
        .map_or(Ok(()), Err)
}
```

Reply:
Thanks, please submit this as another PR so this PR makes the merge freeze.